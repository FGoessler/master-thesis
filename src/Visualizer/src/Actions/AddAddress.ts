import { Action } from "redux";

export interface AddAddress extends Action {
  type: string;
  address: string;
}

export function isAddAddress(action: Action): action is AddAddress {
  return action.type === "AddAddress";
}

export const addAddress = (address: string): AddAddress => {
  return {
    type: "AddAddress",
    address: address
  };
};
