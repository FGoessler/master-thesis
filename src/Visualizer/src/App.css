
body {
    margin: 0;
    padding: 0;
    font-family: "Helvetica Neue", Helvetica, Roboto, Arial, sans-serif;
}

.hash {
    font-family: monospace;
}

.highlight {
    font-weight: bold;
}

.full-height {
    height: 100%;
    overflow: scroll;
}

.detailsTable {
    width: 100%;
}
