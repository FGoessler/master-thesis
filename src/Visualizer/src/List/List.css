
.list {
    background-color: #f1ede4;
    overflow-x: hidden;
    overflow-y: scroll;
    height: 100%;
    border-right: 1px lightgrey solid;
}

.list ul {
    list-style-type: none;
    padding: 0;
}

.list-item {
    cursor: pointer;
    padding: 5px;
}

.list-item:hover {
    background-color: #f7d87c;
}
