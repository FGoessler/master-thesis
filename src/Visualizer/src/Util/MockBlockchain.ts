
import { Block, Eth, Web3, Transaction, TransactionReceipt } from "./web3";
import { BigNumber, BigNumberHelper } from "./bigNumber";

export class MockBlock implements Block {
  number: number = 1;
  hash: string = "0x2a";
  parentHash: string = "0x1a";
  timestamp: number = 100000;
  transactions: (string | Transaction)[] = [];

  constructor(num: number = 1) {
    this.number = num;
  }
}

export class MockTransaction implements Transaction {
  hash: string = "0xffff1a";
  from: string = "0xdd1a";
  to: string  = "0xdd2b";
  value: BigNumber = BigNumberHelper.toBigNumber(0);
  input: string = "";
  transactionReceipt?: TransactionReceipt;
  isContractDeployment: boolean = false;
}

export default class MockWeb3 implements Web3 {

  public eth: Eth;

  constructor() {
    this.eth = new MockEth();
  }
}

export class MockEth implements Eth {

  public transactionReceipts: { [key: string]: TransactionReceipt };
  public blocks: Block[];
  public balances: {};

  constructor() {
    this.transactionReceipts = {
      "0xffff1a": {
        blockHash: "0x2a",
        blockNumber: 1,
        transactionHash: "0xffff1a",
        gasUsed: 100000,
        logs: []
      }
    };

    this.blocks = [
      {
        number: 0,
        hash: "0x1a",
        parentHash: "",
        timestamp: 100000,
        transactions: []
      },
      {
        number: 1,
        hash: "0x2a",
        parentHash: "0x1a",
        timestamp: 100000,
        transactions: [
          new MockTransaction()
        ]
      },
      {
        number: 2,
        hash: "0x3a",
        parentHash: "0x2a",
        timestamp: 100000,
        transactions: []
      }
    ];

    this.balances = {
      "0x1234": BigNumberHelper.toBigNumber("9900000")
    };
  }

  contract(abi: any[]) {
    return {};
  }

  getBlock(hash: number | string, inclTransactions: boolean, cb: (err?: Error, b?: Block) => void): void {
    if (typeof hash === "number") {
      cb(undefined, this.blocks[hash]);
    } else {
      if (hash === "latest") {
        cb(undefined, this.blocks[this.blocks.length - 1]);
      } else {
        const foundBlock = this.blocks.find((b) => { return b.hash === hash; });
        if (foundBlock) {
          cb(undefined, foundBlock);
        } else {
          cb(new Error(`Block ${hash} not found!`), undefined);
        }
      }
    }
  }

  getTransactionReceipt(hash: string, cb: (err?: Error, b?: TransactionReceipt) => void): void {
    if (this.transactionReceipts[hash]) {
      cb(undefined, this.transactionReceipts[hash]);
    } else {
      cb(new Error(`Transaction ${hash} not found!`), undefined);
    }
  }

  getBalance(address: string, cb: ((err?: Error, b?: BigNumber) => void)): void;
  getBalance(address: string, block: number | string, cb: (err?: Error, b?: BigNumber) => void): void;
  getBalance(address: string,
             blockOrCb: number | string | ((err?: Error, b?: BigNumber) => void),
             cb?: ((err?: Error, b?: BigNumber) => void)): void {
    let callback: (err?: Error, b?: BigNumber) => void;
    if (typeof blockOrCb === "function") {
      callback = blockOrCb;
    } else if (cb && typeof cb === "function") {
      callback = cb;
    } else {
      throw new Error("Missing callback!");
    }

    const balance = this.balances[address];
    if (balance) {
      callback(undefined, balance);
    } else {
      callback(new Error(`Address ${address} not found!`), undefined);
    }
  }
}
