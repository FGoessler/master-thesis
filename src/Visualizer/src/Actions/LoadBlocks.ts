import { Block, isTransaction } from "../Util/web3";
import BlockListService from "../BlockInspector/BlockListService";
import { Action, Dispatch } from "redux";
import { addAddress } from "./AddAddress";

export interface LoadingBlocks extends Action {
  type: string;
  loading: boolean;
}

export function isLoadingBlocks(action: Action): action is LoadingBlocks {
  return action.type === "LoadingBlocks";
}

export const loadingBlocks = (loading: boolean): LoadingBlocks => {
  return {
    type: "LoadingBlocks",
    loading: loading
  };
};

export interface LoadedBlocks extends Action {
  type: string;
  blocks: Block[];
  latestBlockNr: number;
  oldestBlockNr: number;
}

export function isLoadedBlocks(action: Action): action is LoadedBlocks {
  return action.type === "LoadedBlocks";
}

export const loadedBlocks = (blocks: Block[], latestBlockNr: number, oldestBlockNr: number): LoadedBlocks => {
  return {
    type: "LoadedBlocks", blocks, latestBlockNr, oldestBlockNr
  };
};

export function loadBlocks(startingBlockNr?: number) {
  return function (dispatch: Dispatch<{}>, _: {}, dependencies: {blockListService: BlockListService}) {

    dispatch(loadingBlocks(true));

    const addAllAddressesInTransactionsOfBlocks = (blocks: Block[]) => {
      blocks.forEach((block) => {
        block.transactions.forEach((transaction) => {
          if (isTransaction(transaction)) {
            dispatch(addAddress(transaction.from));
            if (transaction.to) {
              dispatch(addAddress(transaction.to));
            }
          }
        });
      });
    };
    const onlyBlocksWithTransactionsFilter = (block: Block) => {
      return block.transactions.length > 0;
    };
    const dispatchFilteredLoadedBlocks = (blocks: Block[]) => {
      const filteredBlocks = blocks.filter(onlyBlocksWithTransactionsFilter);
      dispatch(loadedBlocks(
        filteredBlocks,
        blocks.length > 0 ? blocks[0].number : 0,
        blocks.length > 0 ? blocks[blocks.length - 1].number : 0
      ));
      addAllAddressesInTransactionsOfBlocks(filteredBlocks);
    };

    let loadingPromise;
    if (typeof startingBlockNr === "number") {
      loadingPromise = dependencies.blockListService
        .getBlocksStartingWith(startingBlockNr, 3000, dispatchFilteredLoadedBlocks);
    } else {
      loadingPromise = dependencies.blockListService
        .getBlocksStartingWithLatest(3000, dispatchFilteredLoadedBlocks);
    }

    return loadingPromise.then(() => dispatch(loadingBlocks(false)));
  };
}
