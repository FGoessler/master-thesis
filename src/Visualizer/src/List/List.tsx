import * as React from "react";
import Blocklist from "../BlockInspector/BlockList";
import "./List.css";
import { connect, Dispatch } from "react-redux";
import { selectBlock } from "../Actions/LoadBlock";
import { loadBlocks } from "../Actions/LoadBlocks";
import { Block } from "../Util/web3";
import { AppState } from "../App";
import AddressList from "../AddressInspector/AddressList";
import { addAddress } from "../Actions/AddAddress";
import { selectAddress } from "../Actions/SelectAddress";

interface ListProps {
  blocks: Block[];
  latestBlockNr: number;
  oldestBlockNr: number;
  onSelectionChange: (blockHash: string) => void;
  onRequestOlderBlocks: (nextBlock: number) => void;
  loadBlocks: () => void;
  isLoadingBlocks: boolean;
  addresses: string[];
  addressAddHandler: (address: string) => void;
  addressSelectedHandler: (address: string) => void;
}

export class List extends React.Component<ListProps, {}> {

  componentWillMount() {
    if (this.props.loadBlocks) {
      this.props.loadBlocks();
    }
  }

  render() {
    return (
      <div>
        <Blocklist
          blocks={this.props.blocks}
          loading={this.props.isLoadingBlocks}
          latestBlockNr={this.props.latestBlockNr}
          oldestBlockNr={this.props.oldestBlockNr}
          onSelectionChange={this.props.onSelectionChange}
          onRequestOlderBlocks={this.props.onRequestOlderBlocks}
        />
        <AddressList
          addresses={this.props.addresses}
          addressAddHandler={this.props.addressAddHandler}
          addressSelectedHandler={this.props.addressSelectedHandler}
        />
      </div>
    );
  }
}

const mapStateToProps = (state: AppState): Partial<ListProps> => {
  return state.list;
};

const mapDispatchToProps = (dispatch: Dispatch<AppState>): Partial<ListProps> => {
  return {
    onSelectionChange: (blockHash: string) => {
      dispatch(selectBlock(blockHash));
    },
    onRequestOlderBlocks: (nextBlock: number) => {
      dispatch(loadBlocks(nextBlock));
    },
    loadBlocks: () => {
      dispatch(loadBlocks());
    },
    addressAddHandler: (address: string) => {
      dispatch(addAddress(address));
    },
    addressSelectedHandler: (address: string) => {
      dispatch(selectAddress(address));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(List);
