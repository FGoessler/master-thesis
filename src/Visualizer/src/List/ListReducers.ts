import { isLoadedBlocks, isLoadingBlocks } from "../Actions/LoadBlocks";
import { Action, combineReducers } from "redux";
import { Block } from "../Util/web3";
import { isAddAddress } from "../Actions/AddAddress";

export type ListState = {
  blocks: Block[],
  latestBlockNr: number,
  oldestBlockNr: number,
  isLoadingBlocks: boolean,
  addresses: string[]
};

export const blocksReducer = (state: Block[] = [], action: Action): Block[] => {
  if (isLoadedBlocks(action)) {
    return action.blocks;
  } else {
    return state;
  }
};

export const latestBlockReducer = (state: number = 0, action: Action): number => {
  if (isLoadedBlocks(action)) {
    return Math.max(state, action.latestBlockNr);
  } else {
    return state;
  }
};

export const oldestBlockReducer = (state: number = 0, action: Action): number => {
  if (isLoadedBlocks(action)) {
    if (state === 0) {
      return action.oldestBlockNr;
    } else {
      return Math.min(state, action.oldestBlockNr);
    }
  } else {
    return state;
  }
};

export const isLoadingReducer = (state: boolean = false, action: Action): boolean => {
  if (isLoadingBlocks(action)) {
    return action.loading;
  } else {
    return state;
  }
};

export const addAddressReducer = (state: string[] = [], action: Action): string[] => {
  if (isAddAddress(action)) {
    if (state.indexOf(action.address) === -1) {
      return state.concat(action.address);
    } else {
      return state;
    }
  } else {
    return state;
  }
};

export const listReducer = combineReducers({
  blocks: blocksReducer,
  oldestBlockNr: oldestBlockReducer,
  latestBlockNr: latestBlockReducer,
  isLoadingBlocks: isLoadingReducer,
  addresses: addAddressReducer
});
