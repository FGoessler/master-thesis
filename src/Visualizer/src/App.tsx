import * as React from "react";
import "./App.css";
import List from "./List/List";
import Details from "./Details/Details";
import { ListState } from "./List/ListReducers";
import { DetailsState } from "./Details/DetailsReducers";

export interface AppState {
  details: DetailsState;
  list: ListState;
}

export default class App extends React.Component<{}, {}> {

  render() {
    return (
      <div className="container-fluid" style={{height: "100vh"}}>
        <div className="row full-height">
          <div className="col-xs-4 list full-height">
            <List />
          </div>
          <div className="col-xs-8 details full-height">
            <Details />
          </div>
        </div>
      </div>
    );
  }
}
