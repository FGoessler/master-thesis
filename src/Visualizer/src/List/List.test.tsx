import * as React from "react";
import * as ReactDOM from "react-dom";
import configureMockStore, { MockStore } from "redux-mock-store";
import List from "./List";
import { Provider } from "react-redux";
import { MockBlock } from "../Util/MockBlockchain";
import { loadBlocks } from "../Actions/LoadBlocks";
import Mock = jest.Mock;
import { AppState } from "../App";

let mockStore: MockStore<{}>;

beforeEach(() => {
  const initialState: Partial<AppState> = {
    list: {
      blocks: [new MockBlock(1), new MockBlock(2)],
      latestBlockNr: 1,
      oldestBlockNr: 0,
      isLoadingBlocks: false,
      addresses: ["0x1234567890"]
    }
  };
  mockStore = configureMockStore()(initialState);
  mockStore.dispatch = jest.fn();
});

it("renders correctly", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><List/></Provider>, div);
  expect(div).toMatchSnapshot();
});

it("renders loading state correctly", () => {
  const initialState: Partial<AppState> = {
    list: {
      blocks: [],
      latestBlockNr: 0,
      oldestBlockNr: 0,
      isLoadingBlocks: true,
      addresses: []
    }
  };
  mockStore = configureMockStore()(initialState);
  mockStore.dispatch = jest.fn();

  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><List/></Provider>, div);
  expect(div).toMatchSnapshot();
});

it("triggers block load on mount", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><List/></Provider>, div);
  expect((mockStore.dispatch as Mock<Function>).mock.calls[0][0].toString()).toEqual(loadBlocks().toString());
});
