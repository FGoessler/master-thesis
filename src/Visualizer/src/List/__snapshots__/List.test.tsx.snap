// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <div>
      <h3>
        Blocks
      </h3>
      <p>
        (Only showing blocks with transactions)
      </p>
      <ul>
        <li
          class="list-item"
        >
          <!-- react-text: 7 -->
          1
          <!-- /react-text -->
          <!-- react-text: 8 -->
          - 
          <!-- /react-text -->
          <span
            class="hash"
          >
            0x2a
          </span>
        </li>
        <li
          class="list-item"
        >
          <!-- react-text: 11 -->
          2
          <!-- /react-text -->
          <!-- react-text: 12 -->
          - 
          <!-- /react-text -->
          <span
            class="hash"
          >
            0x2a
          </span>
        </li>
      </ul>
      <p>
        <!-- react-text: 15 -->
        Loaded blocks 
        <!-- /react-text -->
        <!-- react-text: 16 -->
        1
        <!-- /react-text -->
        <!-- react-text: 17 -->
        to 
        <!-- /react-text -->
        <!-- react-text: 18 -->
        0
        <!-- /react-text -->
        <!-- react-text: 19 -->
        .
        <!-- /react-text -->
      </p>
      <button>
        Load older blocks
      </button>
    </div>
    <div>
      <h3>
        Addresses
      </h3>
      <input
        type="text"
        placeholder="Add address"
        value=""
      />
      <button>
        Add
      </button>
      <ul>
        <li
          class="list-item hash"
        >
          0x1234567890
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`renders loading state correctly 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <div>
      <h3>
        Blocks
      </h3>
      <p>
        (Only showing blocks with transactions)
      </p>
      <span>
        Loading blocks...
      </span>
      <ul />
      <p>
        <!-- react-text: 8 -->
        Loaded blocks 
        <!-- /react-text -->
        <!-- react-text: 9 -->
        0
        <!-- /react-text -->
        <!-- react-text: 10 -->
        to 
        <!-- /react-text -->
        <!-- react-text: 11 -->
        0
        <!-- /react-text -->
        <!-- react-text: 12 -->
        .
        <!-- /react-text -->
      </p>
      <button>
        Load older blocks
      </button>
    </div>
    <div>
      <h3>
        Addresses
      </h3>
      <input
        type="text"
        placeholder="Add address"
        value=""
      />
      <button>
        Add
      </button>
      <ul />
    </div>
  </div>
</div>
`;
